source/opt/loop_fusion.cpp

```diff
@@ -0,0 +1,735 @@
+    return false;
+  }
+
+  for (uint32_t i = 0; i < condition_0->NumInOperandWords(); ++i) {
+    auto arg_0 = context_->get_def_use_mgr()->GetDef(
+        condition_0->GetSingleWordInOperand(i));
+    auto arg_1 = context_->get_def_use_mgr()->GetDef(
+        condition_1->GetSingleWordInOperand(i));
+
+    if (arg_0 == induction_0_ && arg_1 == induction_1_) {
+      continue;
+    }
+
+    if (arg_0 == induction_0_ && arg_1 != induction_1_) {
+      return false;
+    }
+
+    if (arg_1 == induction_1_ && arg_0 != induction_0_) {
+      return false;
+    }
+
+    if (arg_0 != arg_1) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool LoopFusion::CheckStep() {
+  auto scalar_analysis = context_->GetScalarEvolutionAnalysis();
+  SENode* induction_node_0 = scalar_analysis->SimplifyExpression(
+      scalar_analysis->AnalyzeInstruction(induction_0_));
+  if (!induction_node_0->AsSERecurrentNode()) {
+    return false;
+  }
+
+  SENode* induction_step_0 =
+      induction_node_0->AsSERecurrentNode()->GetCoefficient();
+  if (!induction_step_0->AsSEConstantNode()) {
+    return false;
+  }
+
+  SENode* induction_node_1 = scalar_analysis->SimplifyExpression(
+      scalar_analysis->AnalyzeInstruction(induction_1_));
+  if (!induction_node_1->AsSERecurrentNode()) {
+    return false;
+  }
+
+  SENode* induction_step_1 =
+      induction_node_1->AsSERecurrentNode()->GetCoefficient();
+  if (!induction_step_1->AsSEConstantNode()) {
+    return false;
+  }
+
+  if (*induction_step_0 != *induction_step_1) {
+    return false;
+  }
+
+  return true;
+}
+
+std::map<ir::Instruction*, std::vector<ir::Instruction*>>
+LoopFusion::LocationToMemOps(const std::vector<ir::Instruction*>& mem_ops) {
+  std::map<ir::Instruction*, std::vector<ir::Instruction*>> location_map{};
+
+  for (auto instruction : mem_ops) {
+    auto access_location = context_->get_def_use_mgr()->GetDef(
+        instruction->GetSingleWordInOperand(0));
+
+    if (access_location->opcode() == SpvOpAccessChain) {
+      // get array
+      auto access_array = context_->get_def_use_mgr()->GetDef(
+          access_location->GetSingleWordInOperand(0));
+
+      location_map[access_array].push_back(instruction);
+    } else {
+      location_map[access_location].push_back(instruction);
+    }
+  }
+
+  return location_map;
+}
+
+std::vector<ir::Instruction*> LoopFusion::GetInstructionsInLoopForOpcode(
+    ir::Loop* loop, SpvOp opcode) {
+  std::vector<ir::Instruction*> instructions{};
+
+  for (auto block_id : loop->GetBlocks()) {
+    if (block_id == loop->GetLatchBlock()->id()) {
+      continue;
+    }
+
+    for (auto& instruction : *containing_function_->FindBlock(block_id)) {
+      if (instruction.opcode() == opcode) {
+        instructions.push_back(&instruction);
+      }
+    }
+  }
+
+  return instructions;
+}
+
+bool LoopFusion::IsUsedInLoop(ir::Instruction* instruction, ir::Loop* loop) {
+  auto used = false;
+
+  context_->get_def_use_mgr()->ForEachUser(
```

Comment:
You can use `WhileEachUser`  to get an early exit when you found a use inside the loop.